openml/datasets/dataset.py

```diff
@@ -295,17 +372,73 @@ def _data_arff_to_pickle(self, data_file):
                 else:
                     col.append(X[column_name])
             X = pd.concat(col, axis=1)
+        else:
+            raise ValueError("Dataset format '{}' is not a valid format.".format(self.format))
+
+        return X, categorical, attribute_names
+
+    def _create_pickle_in_cache(self, data_file: str) -> str:
+        """ Parse the arff and pickle the result. Update any old pickle objects. """
+        data_pickle_file = data_file.replace('.arff', '.pkl.py3')
+        if os.path.exists(data_pickle_file):
+            # Load the data to check if the pickle file is outdated (i.e. contains numpy array)
+            with open(data_pickle_file, "rb") as fh:
+                try:
+                    data, categorical, attribute_names = pickle.load(fh)
+                except EOFError:
+                    # The file is likely corrupt, see #780.
+                    # We deal with this when loading the data in `_load_data_from_disk`.
+                    return data_pickle_file
+
+            # Between v0.8 and v0.9 the format of pickled data changed from
+            # np.ndarray to pd.DataFrame. This breaks some backwards compatibility,
+            # e.g. for `run_model_on_task`. If a local file still exists with
+            # np.ndarray data, we reprocess the data file to store a pickled
+            # pd.DataFrame blob. See also #646.
+            if isinstance(data, pd.DataFrame) or scipy.sparse.issparse(data):
+                logger.debug("Data pickle file already exists and is up to date.")
+                return data_pickle_file
+
+        # At this point either the pickle file does not exist, or it had outdated formatting.
+        # We parse the data from arff again and populate the cache with a recent pickle file.
+        X, categorical, attribute_names = self._parse_data_from_arff(data_file)
 
-        # Pickle the dataframe or the sparse matrix.
         with open(data_pickle_file, "wb") as fh:
-            pickle.dump((X, categorical, attribute_names), fh, -1)
+            pickle.dump((X, categorical, attribute_names), fh, pickle.HIGHEST_PROTOCOL)
         logger.debug("Saved dataset {did}: {name} to file {path}"
                      .format(did=int(self.dataset_id or -1),
                              name=self.name,
                              path=data_pickle_file)
                      )
+
         return data_pickle_file
 
+    def _load_data(self):
+        """ Load data from pickle or arff. Download data first if not present on disk. """
+        if self.data_pickle_file is None:
+            if self.data_file is None:
+                self._download_data()
+            self.data_pickle_file = self._create_pickle_in_cache(self.data_file)
+
+        try:
+            with open(self.data_pickle_file, "rb") as fh:
+                data, categorical, attribute_names = pickle.load(fh)
+        except EOFError:
+            warn(
```

Comment:
Could you please use the logger here?

Reply:
Had that initially but I saw `warn` used in the format conversion function, and figured this is more important, so adopted `warn`. Will change to log.